site/content/blog/argocd-carvel-plugin.md

```diff
@@ -0,0 +1,159 @@
+
+```yaml
+#! repo-server-overlay.yml
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.subset({"kind":"Deployment", "metadata":{"name":"argocd-repo-server"}}), expects=1
+---
+spec:
+  template:
+    spec:
+      #! 1. Define an emptyDir volume which will hold the custom binaries
+      volumes:
+        - name: custom-tools
+          emptyDir: {}
+      #! 2. Use an init container to copy custom binaries into the emptyDir
+      initContainers:
+        - name: download-carvel-tools
+          image: index.docker.io/k14s/image@sha256:6ab29951e0207fde6760f6db227f218f20e875f45b22e8ca0ee06c0c8cab32cd
+          command: [sh, -c]
+          args:
+            - cp /usr/local/bin/ytt /custom-tools/ytt
+          volumeMounts:
+            - mountPath: /custom-tools
+              name: custom-tools
+      #! 3. Volume mount the custom binary to the bin directory
+      containers:
+        #@overlay/match by=overlay.subset({"name": "argocd-repo-server"}), expects=1
+        - name: argocd-repo-server
+          volumeMounts:
+            - mountPath: /usr/local/bin/ytt
+              name: custom-tools
+              subPath: ytt
+```
+
+### Make the carvel-ytt plugin available to Applications 
+Expose the plugin by patching the `argocd-cm` ConfigMap's `configManagementPlugins`. The command under `generate` is the command that will be run when using the plugin. 
+
+```yaml
+#! argocd-cm-overlay.yml
+#@ load("@ytt:overlay", "overlay")
+#@overlay/match by=overlay.subset({"kind":"ConfigMap", "metadata":{"name":"argocd-cm"}})
+---
+#@overlay/match missing_ok=True
+data:
+  #! TODO append instead of overwrite this
+  configManagementPlugins: |
+    - name: carvel-ytt
+      generate:                      # Command to generate manifests YAML
+        command: ["sh", "-c"]
+        args: ["ytt -f ."]
+```
+
+### Apply the changes to the cluster
+Since these overlays need to patch the Argo CD configuration, create the namespace and apply the overlays with the Argo CD installation manifests.
+
+```shell
+$ kubectl create namespace argocd
+```
+
+Now, lets install Argo CD Core. To keep setup simple use the core version that does not include the UI.
+
+Run this command to apply the ytt overlays to the Argo CD core manifests and then apply that to the cluster using kapp (or modify for kubectl, if you prefer). The `-y` continues deployment without confirmation. 
+
+```shell
+$ ytt -f argocd-cm-overlay.yml -f  repo-server-overlay.yml -f https://raw.githubusercontent.com/argoproj/argo-cd/v2.2.5/manifests/core-install.yaml  | kapp deploy --app argo --namespace argocd --file - -y
```

Comment:
lets flip this into `kapp deploy -a ... -f <(ytt ...)` so that we dont have to specify -y. this gives user opportunity to review and decide whether they want to proceed.